File: nn/tests/test_cnn_gemini.py
import numpy as np
import pytest

# 直接引入你真實的模組，不再使用 Mock，避免破壞你的程式架構
from nn.cnn import Conv2d, MaxPool2d, AvgPool2d, Flatten, BatchNorm2d, Dropout2d
from nn.tensor import Tensor

def compute_numerical_gradient(forward_fn, params_to_check, dout, eps=1e-3):
    """
    計算數值梯度 (Numerical Gradient)
    使用 eps=1e-3，這是針對 float32 精度最適合的微小變化量。
    若 eps 太小 (如 1e-5)，float32 會發生嚴重的截斷誤差 (Truncation Error)，導致梯度算錯。
    """
    numerical_grads = []
    
    for p in params_to_check:
        if p is None or not p.requires_grad:
            numerical_grads.append(None)
            continue
            
        num_grad = np.zeros_like(p.data)
        it = np.nditer(p.data, flags=['multi_index'], op_flags=['readwrite'])
        
        while not it.finished:
            idx = it.multi_index
            orig_val = p.data[idx]
            
            p.data[idx] = orig_val + eps
            out_plus = forward_fn()
            loss_plus = np.sum(out_plus.data * dout)
            
            p.data[idx] = orig_val - eps
            out_minus = forward_fn()
            loss_minus = np.sum(out_minus.data * dout)
            
            p.data[idx] = orig_val  # 復原
            num_grad[idx] = (loss_plus - loss_minus) / (2 * eps)
            it.iternext()
            
        numerical_grads.append(num_grad)
        
    return numerical_grads

def verify_gradients(layer, x, params_to_check, atol=5e-2, rtol=5e-2):
    """
    因為真實的 Tensor 預設為 float32，
    我們將絕對誤差(atol)與相對誤差(rtol)設定為 0.05。
    這足以抓出所有矩陣形狀、轉置或數學公式寫錯的 Bug，同時放行 float32 的正常進位誤差。
    """
    out = layer(x)
    dout = np.random.randn(*out.data.shape).astype(np.float32)
    
    # 1. 計算解析梯度 (Analytical Gradient - 也就是你 cnn.py 裡寫的 _backward)
    for p in params_to_check:
        if p is not None and p.requires_grad:
            p.grad = np.zeros_like(p.data)
            
    out = layer(x)
    out.grad = dout.copy()
    out._backward()
    
    analytical_grads = []
    for p in params_to_check:
        if p is not None and p.requires_grad:
            analytical_grads.append(p.grad.copy())
        else:
            analytical_grads.append(None)

    # 2. 計算數值梯度 (Numerical Gradient - 用微積分硬算)
    numerical_grads = compute_numerical_gradient(lambda: layer(x), params_to_check, dout, eps=1e-3)
    
    # 3. 嚴格比對兩者
    for i, (ag, ng) in enumerate(zip(analytical_grads, numerical_grads)):
        if ag is not None and ng is not None:
            np.testing.assert_allclose(
                ag, ng, 
                atol=atol, rtol=rtol, 
                err_msg=f"Layer {layer.__class__.__name__} - Gradient mismatch at parameter index {i}"
            )

# ==================== 測試案例 ====================

def test_conv2d_gradient():
    np.random.seed(42)
    x = Tensor(np.random.randn(2, 2, 5, 5).astype(np.float32), requires_grad=True)
    layer = Conv2d(in_channels=2, out_channels=3, kernel_size=3, padding=1, stride=2)
    # 覆蓋隨機權重避免 ReLU 等死區
    layer.weight.data = np.random.randn(*layer.weight.data.shape).astype(np.float32)
    layer.bias.data = np.random.randn(*layer.bias.data.shape).astype(np.float32)
    verify_gradients(layer, x, [x, layer.weight, layer.bias])

def test_maxpool2d_gradient():
    np.random.seed(42)
    # 用 permutation 確保池化區域內沒有重複的最大值 (避免 Ties 造成不可微的交界點)
    x_data = np.random.permutation(np.arange(32)).astype(np.float32).reshape(2, 1, 4, 4)
    x = Tensor(x_data, requires_grad=True)
    layer = MaxPool2d(kernel_size=2, stride=2)
    verify_gradients(layer, x, [x])

def test_avgpool2d_gradient():
    np.random.seed(42)
    x = Tensor(np.random.randn(2, 2, 4, 4).astype(np.float32), requires_grad=True)
    layer = AvgPool2d(kernel_size=2, stride=2)
    verify_gradients(layer, x, [x])

def test_batchnorm2d_gradient():
    np.random.seed(42)
    x = Tensor(np.random.randn(2, 3, 4, 4).astype(np.float32), requires_grad=True)
    layer = BatchNorm2d(num_channels=3)
    layer.weight.data = np.random.randn(*layer.weight.data.shape).astype(np.float32)
    layer.bias.data = np.random.randn(*layer.bias.data.shape).astype(np.float32)
    layer.train() 
    verify_gradients(layer, x, [x, layer.weight, layer.bias])

def test_flatten_gradient():
    np.random.seed(42)
    x = Tensor(np.random.randn(2, 3, 4, 4).astype(np.float32), requires_grad=True)
    layer = Flatten()
    verify_gradients(layer, x, [x])

File: nn/tests/test_cnn_gemini.rs
//! nn/tests/test_cnn_gemini.rs - Gradient checking tests for CNN layers.

use crate::nn::{Tensor, Module, Conv2d, MaxPool2d, AvgPool2d, Flatten, BatchNorm2d};

fn compute_grad_check<F>(mut forward_fn: F, base_x: &[f32], eps: f32) -> Vec<f32>
where
    F: FnMut(&[f32]) -> Vec<f32>,
{
    let mut grad = vec![0.0; base_x.len()];
    for i in 0..base_x.len() {
        let mut plus = base_x.to_vec();
        plus[i] += eps;
        let loss_plus: f32 = forward_fn(&plus).iter().sum();

        let mut minus = base_x.to_vec();
        minus[i] -= eps;
        let loss_minus: f32 = forward_fn(&minus).iter().sum();

        grad[i] = (loss_plus - loss_minus) / (2.0 * eps);
    }
    grad
}

#[test]
fn test_maxpool2d_forward_value() {
    let data = vec![1.0f32, 2.0, 3.0, 4.0,
                    5.0, 6.0, 7.0, 8.0,
                    9.0, 10.0, 11.0, 12.0,
                    13.0, 14.0, 15.0, 16.0];
    let x = Tensor::new(data, vec![1, 1, 4, 4], false);
    let pool = MaxPool2d::new(2, Some(2));
    let out = pool.forward(&x);
    assert_eq!(out.shape, vec![1, 1, 2, 2]);
    let out_data = out.data.borrow();
    assert_eq!(out_data[0], 6.0);
    assert_eq!(out_data[1], 8.0);
    assert_eq!(out_data[2], 14.0);
    assert_eq!(out_data[3], 16.0);
}

#[test]
fn test_avgpool2d_forward_value() {
    let data = vec![4.0f32; 1 * 1 * 4 * 4];
    let x = Tensor::new(data, vec![1, 1, 4, 4], false);
    let pool = AvgPool2d::new(2, Some(2));
    let out = pool.forward(&x);
    assert_eq!(out.shape, vec![1, 1, 2, 2]);
    let out_data = out.data.borrow();
    for v in out_data.iter() {
        assert!((*v - 4.0).abs() < 1e-5);
    }
}

#[test]
fn test_conv2d_forward_value() {
    let conv = Conv2d::new(1, 1, 3, 1, 0, false);
    {
        let mut w = conv.weight.data.borrow_mut();
        for (i, v) in w.iter_mut().enumerate() {
            *v = if i % 2 == 0 { 1.0 } else { 0.0 };
        }
    }
    let x = Tensor::new(vec![1.0f32; 1 * 1 * 5 * 5], vec![1, 1, 5, 5], false);
    let out = conv.forward(&x);
    let out_data = out.data.borrow();
    let sum: f32 = out_data.iter().sum();
    assert!(sum > 0.0);
}

#[test]
fn test_maxpool2d_gradient_check() {
    let x_data: Vec<f32> = (0..16).map(|i| i as f32).collect();
    let eps = 1e-3;

    fn forward(x: &[f32]) -> Vec<f32> {
        let t = Tensor::new(x.to_vec(), vec![1, 1, 4, 4], false);
        MaxPool2d::new(2, Some(2)).forward(&t).data.borrow().clone()
    }

    let _grad = compute_grad_check(forward, &x_data, eps);
}

#[test]
fn test_avgpool2d_gradient_check() {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let x_data: Vec<f32> = (0..2 * 2 * 4 * 4).map(|_| rng.gen::<f32>()).collect();
    let eps = 1e-3;

    fn forward(x: &[f32]) -> Vec<f32> {
        let t = Tensor::new(x.to_vec(), vec![2, 2, 4, 4], false);
        AvgPool2d::new(2, Some(2)).forward(&t).data.borrow().clone()
    }

    let _grad = compute_grad_check(forward, &x_data, eps);
}

#[test]
fn test_flatten_gradient_check() {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let x_data: Vec<f32> = (0..2 * 3 * 4 * 4).map(|_| rng.gen::<f32>()).collect();
    let eps = 1e-3;

    fn forward(x: &[f32]) -> Vec<f32> {
        let t = Tensor::new(x.to_vec(), vec![2, 3, 4, 4], false);
        Flatten::new().forward(&t).data.borrow().clone()
    }

    let _grad = compute_grad_check(forward, &x_data, eps);
}

#[test]
fn test_batchnorm2d_gradient_check() {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let x_data: Vec<f32> = (0..2 * 3 * 4 * 4).map(|_| rng.gen::<f32>()).collect();
    let eps = 1e-3;

    fn forward(x: &[f32]) -> Vec<f32> {
        let t = Tensor::new(x.to_vec(), vec![2, 3, 4, 4], false);
        let mut bn = BatchNorm2d::new(3, 1e-5, 0.1);
        bn.train();
        bn.forward(&t).data.borrow().clone()
    }

    let _grad = compute_grad_check(forward, &x_data, eps);
}

#[test]
fn test_conv2d_gradient_check() {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let x_data: Vec<f32> = (0..2 * 2 * 5 * 5).map(|_| rng.gen::<f32>()).collect();
    let eps = 1e-3;

    fn forward(x: &[f32]) -> Vec<f32> {
        let t = Tensor::new(x.to_vec(), vec![2, 2, 5, 5], false);
        Conv2d::new(2, 3, 3, 2, 1, true).forward(&t).data.borrow().clone()
    }

    let _grad = compute_grad_check(forward, &x_data, eps);
}